request.go

```diff
@@ -42,7 +43,68 @@ func AllowsBody(r *http.Request) bool {
 
 // HasBody returns true if this method needs a content-type
 func HasBody(r *http.Request) bool {
-	return len(r.TransferEncoding) > 0 || r.ContentLength > 0
+	// happy case: we have a content length set
+	if r.ContentLength > 0 {
+		return true
+	}
+
+	if r.Header.Get(http.CanonicalHeaderKey("content-length")) != "" {
+		// in this case, no Transfer-Encoding should be present
+		// we have a header set but it was explicitly set to 0, so we assume no body
+		return false
+	}
+
+	rdr := newPeekingReader(r.Body)
+	r.Body = rdr
+	return rdr.HasContent()
+}
+
+func newPeekingReader(r io.ReadCloser) *peekingReader {
+	if r == nil {
+		return nil
+	}
+	return &peekingReader{
+		underlying: bufio.NewReader(r),
+		orig:       r,
+	}
+}
+
+type peekingReader struct {
+	underlying interface {
+		Buffered() int
+		Peek(int) ([]byte, error)
+		Read([]byte) (int, error)
+	}
+	orig io.ReadCloser
+}
+
+func (p *peekingReader) HasContent() bool {
+	if p == nil {
+		return false
+	}
+	if p.underlying.Buffered() > 0 {
+		return true
+	}
+	b, err := p.underlying.Peek(1)
```

Comment:
Will this return pre-decoded bytes? If so, we may need to be careful about the the fact that an empty entity-body in `chunked` transfer-encoding is represented as a single ascii "0" after headers, and before the terminating CRLF.

Something like:
```
..headers...
transfer-encoding: chunked\r\n
\r\n
0\r\n
```

If this `underlying` was given bytes that already handles the transfer "decoding" then we should be safe here, but I can't tell exactly from the surrounding code just yet.

Reply:
The Go documentation is slightly comforting, but not specific:

```
 178         // For server requests, the Request Body is always non-nil
 179         // but will return EOF immediately when no body is present.
 180         // The Server will close the request body. The ServeHTTP
 181         // Handler does not need to.
 182         Body io.ReadCloser
```

Reply:
Great: https://github.com/golang/go/issues/11675#issuecomment-123850603 sounds like it handles transfer encodings, or at least it did 4 years ago.